### PR TITLE
mgr/orch: ServiceDescription: Make spec a requirement 

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Because, a `ServiceDescription` is actually  super set of a `ServiceSpec`

## DNM + TODO:

- [x] fix dashboard
- [x] rebase to master

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by liewegas** [comment]
We may want to send this one to master and backport...

**Comment by sebastian-philipp** [comment]
> We may want to send this one to master and backport...

see "TODO"

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1839,18 +1839,25 @@ def describe_service(self, service_type=None, service_name=None,
...
                 if dd.service_name() in self.spec_store.specs:
                     spec = self.spec_store.specs[dd.service_name()]
+                else:
+                    spec = ServiceSpec(
+                        unmanaged=True,
+                        service_type=dd.daemon_type,
+                        service_id=n,
+                        placement=PlacementSpec(
+                            hosts=[dd.hostname]
+                        )
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py]: this is actually questionable. after running

`ceph orch ls  --format yaml | ceph orch apply -i -´ 

this ends up in the store as `unmanged`

**Comment by sebastian-philipp** [comment]
> code (minus the dashboard parts that I don't have a clue of) looks good.
> 
> However, if we drop the `ceph orch spec dump` functionality in favor for `ceph orch ls --format yaml|json|pretty-json` _imo_ we should make
> 
> `ceph orch ls --format yaml | tee servicespecs.yaml && ceph orch apply -i servicespecs.yaml`
> (if you want to preserve the dump file)
> 
> certainly possible to allow importing and exporting of servicespecfiles (for introspection, debugging, environment replication etc)
> 
> The current output however looks like this:
> 
> ```yaml
> placement:
>   host_pattern: '*'
> service_name: crash
> service_type: crash
> status:
>   container_image_id: c07ead29a8ebcac6ead990751693522e6427458477c04b7c0a0e436817a36f50
>   container_image_name: quay.io/ceph-ci/ceph:octopus
>   created: '2020-03-23T14:15:42.063190'
>   last_refresh: '2020-03-23T14:36:05.303556'
>   running: 1
>   size: 1
> ```
> 
> which isn't quite what `ceph orch apply` expects.
> 
> I don't think we should trim or even alter the output of `yaml` but rather add a `--format` parameter like "spec" or so?

I've run `ceph orch ls --format yaml | ceph orch apply -i -` and it worked!

**Comment by jschmid1** [comment]
right, because it will ignore the `status` bits I assume (haven't actually checked if that's the case). But I think we'll reuse the `ceph orch ls --format yaml` to export servicespecs. Manually trimming certain areas is not appreciated I reckon.

**Comment by sebastian-philipp** [comment]
got it.

File: src/ceph.in [hunk 2]
```diff
@@ -333,7 +333,7 @@ def parse_cmdargs(args=None, target=''):
                         help="make less verbose")
 
     parser.add_argument('-f', '--format', choices=['json', 'json-pretty',
-                        'xml', 'xml-pretty', 'plain'], dest='output_format')
+                        'xml', 'xml-pretty', 'plain', 'yaml'], dest='output_format')
```
Comment (liewegas) [inline src/ceph.in]: we should open a tracker item for a YAMLFormatter for common/Formatter.h

File: src/python-common/ceph/deployment/service_spec.py [hunk 3]
```diff
@@ -395,10 +395,22 @@ def from_json(cls, json_spec):
         :param json_spec: A valid dict with ServiceSpec
         """
 
-        service_type = json_spec.get('service_type', '')
+        c = json_spec.copy()
+
+        # kludge to make `from_json` compatible to `Orchestrator.describe_service`
+        # Open question: Remove `service_id` form to_json?
```
Comment (liewegas) [inline src/python-common/ceph/deployment/service_spec.py]: i think yes... if service_name is always present then service_type is redundant

**Comment by liewegas** [comment]
Overall I think this is great!  I think a few additional cleanups would be nice but otherwise looks good

**Comment by jschmid1** [comment]
> got it. (took a while)
> 
> What about [`ceph orch ls --format yaml --export`](https://github.com/kubernetes/kubernetes/issues/24873#issuecomment-215817605) ?

Although this doesn't make a difference for the implementation, I'd swap the flags when documenting it :)

`ceph orch ls --export --format yaml` 

reads more like an English sentence.

File: src/ceph.in (hunk 2 above)
Comment (sebastian-philipp) [reply src/ceph.in]: https://tracker.ceph.com/issues/44745

File: src/python-common/ceph/deployment/service_spec.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/service_spec.py]: hm. rethinking this. I actually like `service_type`, as it clearly states the type of the spec.

**Comment by sebastian-philipp** [comment]
> > got it. (took a while)
> > What about [`ceph orch ls --format yaml --export`](https://github.com/kubernetes/kubernetes/issues/24873#issuecomment-215817605) ?
> 
> Although this doesn't make a difference for the implementation, I'd swap the flags when documenting it :)
> 
> `ceph orch ls --export --format yaml`
> 
> reads more like an English sentence.

done

File: doc/mgr/orchestrator.rst [hunk 4]
```diff
@@ -206,18 +206,29 @@ services of a particular type via optional --type parameter
...
+
+    ceph orch ls --export
+
 
+Daemon Status
+=============
 
+Print a list of all daemons known to the orchestrator::
+
+    ceph orch ps [--hostname host] [--daemon_type type] [--daemon_id id] [--format f] [--refresh]
```
Comment (bk201) [inline doc/mgr/orchestrator.rst]:
```suggestion
    ceph orch ps [--hostname host] [--service_name name] [--daemon_type type] [--daemon_id id] [--format f] [--refresh]
```

File: doc/mgr/orchestrator.rst [hunk 5]
```diff
@@ -206,18 +206,29 @@ services of a particular type via optional --type parameter
...
+    
 Query the status of a particular service instance (mon, osd, mds, rgw).  For OSDs
 the id is the numeric OSD ID, for MDS services it is the file system name::
 
-    ceph orch daemon status <type> <instance-name> [--refresh]
+    ceph orch pr --daemon_type osd --daemon_id 0
```
Comment (bk201) [inline doc/mgr/orchestrator.rst]:
```suggestion
    ceph orch ps --daemon_type osd --daemon_id 0
```

File: src/pybind/mgr/orchestrator/module.py [hunk 6]
```diff
@@ -24,14 +24,25 @@
...
     else:
         return '-'
 
+
+def to_format(what, format):
+    if format == 'json':
+        return json.dumps(what, sort_keys=True)
+    elif format == 'json-pretty':
+        return json.dumps(what, indent=2, sort_keys=True)
+    elif format == 'yaml':
```
Comment (bk201) [inline src/pybind/mgr/orchestrator/module.py]:
```
bin/ceph orch ls --format=yaml       
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
usage: ceph [-h] [-c CEPHCONF] [-i INPUT_FILE] [-o OUTPUT_FILE] [--setuser SETUSER] [--setgroup SETGROUP] [--id CLIENT_ID] [--name CLIENT_NAME] [--cluster CLUSTER]
            [--admin-daemon ADMIN_SOCKET] [-s] [-w] [--watch-debug] [--watch-info] [--watch-sec] [--watch-warn] [--watch-error] [-W WATCH_CHANNEL] [--version] [--verbose]
            [--concise] [-f {json,json-pretty,xml,xml-pretty,plain}] [--connect-timeout CLUSTER_TIMEOUT] [--block] [--period PERIOD]
ceph: error: argument -f/--format: invalid choice: 'yaml' (choose from 'json', 'json-pretty', 'xml', 'xml-pretty', 'plain')
```
Hmm..seems guarding in cpp code.

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/module.py]: https://github.com/ceph/ceph/pull/34061/files#diff-b99cad6b446ed3108100f38e6b6249d8R336

you'll need to use the correct ceph binary

**Comment by sebastian-philipp** [comment]
rebased

**Comment by muhlba91** [comment]
any info when this fix will be available in an octopus release? :)

**Comment by sebastian-philipp** [comment]
probably 15.2.2

**Comment by muhlba91** [comment]
thank you. what's the release schedule to expect 15.2.2 then? :)

**Comment by lbogdan** [comment]
We're at `15.2.3` and this still happens. Any timeline for a release including the fix?

**Comment by lbogdan** [comment]
FYI: This was fixed with the latest `ceph/ceph:v15.2.4` image.

**Pull request merged**